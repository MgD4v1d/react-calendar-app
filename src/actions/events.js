import Swal from "sweetalert2";
import { fetchConToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";


export const startAddNew = ( event ) => {
    return async( dispatch, getState ) => {

        const {uid, name} = getState().auth;

        try {
            const resp = await fetchConToken('events/new', event, 'POST')
            const body = await resp.json();

            if(body.ok){

                event.id = body.evento.id;
                event.user = {
                    _id: uid,
                    name: name
                }

                dispatch(eventAddNew(event));
            }

        } catch (error) {
            console.log(error)
        }

    }
}



export const startLoadingEvents = () => {
    return async (dispatch) => {

        try{

            const resp = await fetchConToken('events');
            const body = await resp.json();

            const events = prepareEvents( body.eventos );

            if(body.ok){
                dispatch(eventLoaded(events));
            }


        }catch(error){
            console.log(error);
        }

    }
}


export const eventStartUpdate = ( event ) => {
    return async (dispatch) => {

        try {
            
            const resp = await fetchConToken(`events/edit/${ event.id }`, event, 'PUT');
            const body = await resp.json();

            if(body.ok){
                dispatch(eventUpdated(event));
            }else{
                Swal.fire('Error', body.msg , 'error');
            }


        } catch (error) {
            console.log(error)
        }

    }
}


export const eventStartDelete = () => {
    return async (dispatch, getState) => {

        const { id } = getState().calendar.activeEvent;


        try {

            const resp = await fetchConToken(`events/delete/${id}`, {}, 'DELETE');
            const body = await resp.json();

            if(body.ok){
                dispatch(eventDeleted());
            }else{
                Swal.fire('Error', body.msg , 'error');
            }

            
        } catch (error) {
            console.log(error);
        }

    }
}


const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});


export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});


export const eventClearActiveEvent = () => ({
    type: types.eventClearActiveEvent,
});


export const eventLogoutClean = () => ({
    type: types.eventLogoutCleaning,
})

const eventUpdated = (event) => ({
    type: types.eventUpdated,
    payload: event
});

const eventDeleted = () => ({
    type: types.eventDeleted,
})


const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
})