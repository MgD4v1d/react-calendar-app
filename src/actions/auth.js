import Swal from "sweetalert2";
import { fetchConToken, fetchSinToken } from "../helpers/fetch";
import { types } from "../types/types";
import { eventLogoutClean } from "./events";


export const startLogin = (email, password) =>{

    return async(dispatch) => {

        const resp = await fetchSinToken('auth', { email, password }, 'POST');

        const body = await resp.json();

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch(login({
                uid: body.uid,
                name: body.name
            }));

        }else {
            Swal.fire('Error', body.msg, 'error');
        }

    }
}


export const startRegister = (name, email, password) => {
    return async(dispatch) => {


        const resp = await fetchSinToken('auth/register', { name, email, password }, 'POST');
        const body = await resp.json();

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch(login({
                uid: body.uid,
                name: body.name
            }));

        }else{
            Swal.fire('Error', body.msg, 'error')
        }


    }
}


export const startChecking = () => {
    return async(dispatch) => {

        const resp = await fetchConToken('auth/renew');
        const body = await resp.json();

        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch(login({
                uid: body.uid,
                name: body.name
            }));

        }else {
            
            dispatch(checkinFinish());
        }

    }
}


export const startLogout = () => {
    return (dispatch) => {

        localStorage.clear();
        dispatch(eventLogoutClean());
        dispatch(logout());

    }
}

const checkinFinish = () => ({
    type: types.authCheckingFinish,
});

const login = (user) => ({
    type: types.authLogin,
    payload: user
});


const logout = () => ({
    type: types.authLogout
});

