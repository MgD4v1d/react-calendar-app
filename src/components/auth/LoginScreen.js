import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogin } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import validator from 'validator';

import './login.css';
import Swal from 'sweetalert2';

export const LoginScreen = () => {

  const dispatch = useDispatch();

  const [inputValues, handleInputChange] = useForm({
    email: '',
    password: ''
  });


  const { email, password } = inputValues;

  const handleLogin = (e) => {
    e.preventDefault();

    if(isValidForm()){
      dispatch(startLogin(email, password));
    }


  }


  const isValidForm = () => {
    if(!validator.isEmail(email)){
      Swal.fire('Error', 'Correo no valido', 'error');
      return false;
    }else if(password.length ===0){
      Swal.fire('Error', 'Contraseña no valida', 'error');
      return false;
    }
    return true;
  }

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-12 login-form-1">
          <h3>Login Calendar</h3>
          <form onSubmit={ handleLogin }>
            <div className="form-group mb-3">
              <input
                type="text"
                className="form-control"
                placeholder="Correo"
                name="email"
                value={email}
                onChange={ handleInputChange }
              />
            </div>
            <div className="form-group mb-3">
              <input
                type="password"
                className="form-control"
                placeholder="Contraseña"
                name="password"
                value={password}
                onChange={ handleInputChange }
              />
            </div>
            <div className="form-group mt-4 d-grid gap-2">
              <input type="submit" className="btn btnSubmit" value="Login" />
            </div>
          </form>
          <Link
            className="link-secondary text-decoration-none"
            to="/auth/register"
          >
              Create a new account
          </Link>
        </div>
      </div>
    </div>
  );
};
