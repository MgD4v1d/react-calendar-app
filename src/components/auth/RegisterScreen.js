import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import Swal from 'sweetalert2';
import { startRegister } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import validator from 'validator';

import "./login.css";

export const RegisterScreen = () => {


  const dispatch = useDispatch();

  const [inputValues, handleInputChange] = useForm({
    name: '',
    email: '',
    password: '',
    password2: ''
  });


  const {name, email, password, password2} = inputValues;


  const handleRegister = (e) => {
    e.preventDefault();

    if(isValidRegisterForm()){
      dispatch(startRegister(name, email, password));
    }


  }


  const isValidRegisterForm = () => {

    if(validator.isEmpty(name)){
      Swal.fire('Error', 'El campo "Nombre" es requerido', 'error');
      return false;
    }else if(name.length <= 2){
      Swal.fire('Error', 'El nombre debe ser de almenos 3 caracteres', 'error');
      return false;
    }else if( !validator.isEmail(email)){
      Swal.fire('Error', 'El correo no tiene un formato valido', 'error');
      return false;
    }else if(password.length <= 2){
      Swal.fire('Error', 'La contraseña al menos debe terner 6 caracteres', 'error');
      return false;
    }else if(password !== password2){
      Swal.fire('Error', 'Las contraseñas no coinciden', 'error');
      return false;
    }
    return true
  }

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-12 login-form-2">
          <h3>Registro</h3>
          <form onSubmit={handleRegister}>
            <div className="form-group mb-3">
              <input
                type="text"
                className="form-control"
                placeholder="Nombre"
                name="name"
                value={name}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group mb-3">
              <input
                type="email"
                className="form-control mb-3"
                placeholder="Correo"
                name="email"
                value={email}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group mb-3">
              <input
                type="password"
                className="form-control"
                placeholder="Contraseña"
                name="password"
                value={password}
                onChange={handleInputChange}
              />
            </div>

            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Repita la contraseña"
                name="password2"
                value={password2}
                onChange={handleInputChange}
              />
            </div>

            <div className=" mt-4 d-grid gap-2">
              <button type="submit" className="btn btn-primary btnSubmit">
                Crear Cuenta
              </button>
            </div>
          </form>
          <Link
          className="text-white text-decoration-none"
          to="/auth/login"
          >
              ¿Ya tienes una cuenta?
          </Link>
        </div>
      </div>
    </div>
  );
};
