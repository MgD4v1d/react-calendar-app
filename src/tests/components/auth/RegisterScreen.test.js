import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { MemoryRouter } from 'react-router-dom';

import Swal from 'sweetalert2';

import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { startRegister } from '../../../actions/auth';




jest.mock('../../../actions/auth',()=>({
    startRegister: jest.fn()
}));

jest.mock('sweetalert2', () =>({
    fire: jest.fn(),
}))

const middlewares= [thunk];

const mockStore = configureStore(middlewares);

const initState = {};

const store = mockStore (initState);

store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
)



describe('Pruebas en el <RegisterScreen />', () => {


    beforeEach(() => {
        jest.clearAllMocks();
    })
    

    test('debe hacer match con el snapshot', () => {
        
        expect(wrapper).toMatchSnapshot();

    });


    test('No hay registro si las contrase単as son diferentes', () => {

        wrapper.find('input[name="email"]').simulate('change', {
            target:{
                name: 'name',
                value: 'Peter'
            }
        });

        wrapper.find('input[name="email"]').simulate('change', {
            target:{
                name: 'email',
                value: 'example@example.com'
            }
        });

        wrapper.find('input[name="password"]').simulate('change', {
            target:{
                name: 'password',
                value: '1234567'
            }
        });

        wrapper.find('input[name="password2"]').simulate('change', {
            target:{
                name: 'password2',
                value: '123456'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        
        expect(Swal.fire).toHaveBeenCalledWith('Error', 'Las contrase単as no coinciden', 'error');
        expect(startRegister).not.toHaveBeenCalled();

    });

    test('debe dispararse el registro con contrase単as iguales', () => {
        
        wrapper.find('input[name="email"]').simulate('change', {
            target:{
                name: 'name',
                value: 'Peter'
            }
        });

        wrapper.find('input[name="email"]').simulate('change', {
            target:{
                name: 'email',
                value: 'example@example.com'
            }
        });

        wrapper.find('input[name="password"]').simulate('change', {
            target:{
                name: 'password',
                value: '123456'
            }
        });

        wrapper.find('input[name="password2"]').simulate('change', {
            target:{
                name: 'password2',
                value: '123456'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect(Swal.fire).not.toHaveBeenCalledWith('Error', 'Las contrase単as no coinciden', 'error');
        expect(startRegister).toHaveBeenCalled();

    });

});