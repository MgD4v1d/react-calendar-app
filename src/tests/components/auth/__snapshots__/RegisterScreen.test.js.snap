// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <RegisterScreen /> debe hacer match con el snapshot 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-12 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group mb-3"
        >
          <input
            className="form-control"
            name="name"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group mb-3"
        >
          <input
            className="form-control mb-3"
            name="email"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group mb-3"
        >
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="password2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className=" mt-4 d-grid gap-2"
        >
          <button
            className="btn btn-primary btnSubmit"
            type="submit"
          >
            Crear Cuenta
          </button>
        </div>
      </form>
      <a
        className="text-white text-decoration-none"
        href="/auth/login"
        onClick={[Function]}
      >
        ¿Ya tienes una cuenta?
      </a>
    </div>
  </div>
</div>
`;
